explore/matrix-io/02.sparse-and-dense.ipynb

```diff
@@ -0,0 +1,427 @@
+    "\n",
+    "Uncompressed sparse matrices load and multiply about as quickly as dense matrices. The order of matrix reads does not change for sparse reads from npz files. Compressed sparse matrices take an order of magnitude more time to load and multiply."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "CCpGeAeGaD\n",
+      "density: 97.8 %\n",
+      "Dense No Memmap: 0.000765 sec\n",
+      "Compressed Sparse: 0.0154 sec\n",
+      "Uncompressed Sparse: 0.00336 sec\n",
+      "\n",
+      "DaGpBP\n",
+      "density: 21.0 %\n",
+      "Dense No Memmap: 0.00262 sec\n",
+      "Compressed Sparse: 0.0177 sec\n",
+      "Uncompressed Sparse: 0.00536 sec\n",
+      "\n",
+      "MFpGdCcSE\n",
+      "density: 14.1 %\n",
+      "Dense No Memmap: 0.0431 sec\n",
+      "Compressed Sparse: 0.104 sec\n",
+      "Uncompressed Sparse: 0.0263 sec\n",
+      "\n",
+      "GiGpBP\n",
+      "density: 4.65 %\n",
+      "Dense No Memmap: 0.556 sec\n",
+      "Compressed Sparse: 0.509 sec\n",
+      "Uncompressed Sparse: 0.144 sec\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "for i in ['CCpGeAeGaD', 'DaGpBP', 'MFpGdCcSE', 'GiGpBP']:\n",
+    "    print(i)\n",
+    "    mat = np.load(f'data/{i}-dense-uncompressed.npy')\n",
+    "    \n",
+    "    density = 100 * (mat != 0).sum() / np.prod(mat.shape)\n",
+    "    print(f'density: {density :.3} %')\n",
+    "\n",
+    "    # Create a vector to multiply\n",
+    "    vector_size = mat.shape[0]\n",
+    "    vec = np.zeros((1, vector_size))\n",
+    "    # 4 search nodes\n",
+    "    indices = np.random.randint(0, high=vector_size, size=4)\n",
+    "    vec[0, indices] = 1\n",
+    "    del mat\n",
+    "    \n",
+    "    t1 = time.time()\n",
+    "    matrix = np.load(f'data/{i}-dense-uncompressed.npy')\n",
+    "    output = vec @ matrix\n",
+    "    t2 = time.time()\n",
+    "    time_1 = t2 - t1\n",
+    "    del matrix\n",
+    "    del output\n",
+    "    \n",
+    "    t1 = time.time()\n",
+    "    matrix = sparse.load_npz(f'data/{i}-sparse-compressed.npz')\n",
+    "    output = vec @ matrix\n",
+    "    t2 = time.time()\n",
+    "    time_2 = t2 - t1\n",
+    "    del matrix\n",
+    "    del output\n",
+    "    \n",
+    "    t1 = time.time()\n",
+    "    matrix = sparse.load_npz(f'data/{i}-sparse-uncompressed.npz')\n",
+    "    output = vec @ matrix\n",
+    "    t2 = time.time()\n",
+    "    time_3 = t2 - t1\n",
+    "    del matrix\n",
+    "    del output\n",
+    "    \n",
+    "    print(f'Dense No Memmap: {time_1 :.3} sec\\nCompressed Sparse: {time_2 :.3} sec\\nUncompressed Sparse: {time_3 :.3} sec\\n')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "collapsed": true
+   },
+   "source": [
+    "## Subset matrix memmap based on nonzero rows of the search vector"
```

Comment:
Converting to milliseconds would make it easier to eyeball these times. Another options could be to do time relative to `no memmap` (i.e. no memmap time / memmap time or subset time).

Reply:
I changed that, for the first two sets of cells it will show milliseconds. Then, having established the absolute times, I made the final comparison, subsetting memmap for nonzero rows, compare relative to the dense no-memmap condition. Unfortunately, I am on my laptop until the evening and don't have enough memory to run the notebooks, so they display the old outputs at the moment.

Reply:
Great excited for the results later!